Support Interactions

JustCat80 opened the pull request:
Moved to #1280

Commit:
Added interaction event, Added Interaction class (1)

Commit:
Fixed accidental commas

Commit:
Update Constants.js

Commit:
Temp commented out InteractionData to test so far

Commit:
Update Interaction.js

Commit:
Added endpoints

Commit:
Forgot / before urls

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Client.js

Commit:
Update Endpoints.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Interaction return changed

Commit:
Attempt to add Interaction Data

Commit:
Update Interaction.js

Commit:
Removed InteractionData.customId

Commit:
Actually did InteractionData Docs

Commit:
acknowledge is just a promise

Commit:
Defer test

Commit:
editResponse test

Commit:
Update Interaction.js

Commit:
Update Endpoints.js

Commit:
Update Endpoints.js

Commit:
Update Interaction.js

Commit:
editResponse Documentation

Commit:
Added testing for content

Commit:
Removed duplicate endpoints

Commit:
Added createFollowup

Commit:
Fixed acknowledge

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Test log

Commit:
Commented out broken .then statement

Commit:
Changed some docs, added .edit

Commit:
Update Interaction.js

Commit:
edit => editParent, editResponse => edit

Commit:
Removed editInteractionMessage

Commit:
Updated Docs

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Added interaction.delete

Commit:
Update Interaction.js

Commit:
Update Interaction.js

Commit:
Removed dead space between comments and functions

Commit:
Update Interaction.js

Commit:
Added Name to InteractionData

bsian03 reviewed (changes requested):
- Lint
- Missing client methods
- Follow constants styling
- InteractionData does not need to be a class
- Don't assign null for missing fields
- Check if fields aren't equal undefined explicitly like in other structures
- Missing typescript documentation

Also note that Abal has voiced his dislike with interactions, but I do not know how he feels about it now. I have suggested to leave interactions off the table for now unless he gives the green light

Commit:
Added Requested changes

• Linted
• Added createInteractionResponse client method
• Added flags to executeWebhook docs
• started using client methods in Interaction
• No longer assign null for fields
• Now checks fields for undefined explicitly
• Fixed Constants Styling (i think?)
• Removed InteractionData class
• Added Typescript documentation
• Added .vscode to .gitignore

Commit:
Removed _client that i forgot to

Commit:
fixed small docs bit

for some reason .request in interactionCreate is not defined so imma try and fix that but i see nothing wrong

Commit:
Added this._client to all method calls, everything now actually works

Commit:
Fixed Interaction docs

Commit:
added flags to executewebhook for interactions

Commit:
added createInteractionResponse to ts docs

Commit:
Fixed some docs

Commit:
Made interaction functions docs InteractionOptions rather then Record<string, unknown>

Not exactly sure how this works but maybe this im sorry if not eeeeeee

Commit:
Added missing client for member

Commit:
Added forgotten guild object

Commit:
Better docs + notes

Commit:
Updated some docs

Commit:
Added createFollowup options

Commit:
Fixed up some ts docs

Commit:
Fixed delete docs, fixed needing embed when its embeds

Commit:
Moved docs a little

Commit:
Components-only for editParent and edit

Commit:
Dont format allowedMentions because editWebhookMessage does right after

ChampionBuffalo1 commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,248 @@
+    }
+
+    /**
+    * Note: You can **not** use more then 1 initial interaction response per interaction.
```
shouldn't it be "than" instead of "then"

JustCat80 replied:
Yea you are right

Commit:
then => than in interaction docs

JustCat80 committed:
Merge pull request #17 from abalabahaha/dev

Pull dev changes

Commit:
Added forgotten property, fixed docs a little

Commit:
Added forgotten consistent, actually follows styling now

Commit:
Added endpoints, Added consistents

Commit:
Id => ID

Commit:
Add more endpoints

Commit:
Add most client slash command functions

Still missing all Permissions related ones

Commit:
Fixed slash command docs

Commit:
Added most guild slash command functions

JustCat80 committed:
Merge pull request #19 from abalabahaha/dev

Catchup dev updates

JustCat80 committed:
Merge pull request #20 from abalabahaha/dev

Catchup dev updates

Commit:
bulkEditCommands

createCommands => bulkEditCommands

createGuildCommands => bulkEditGuildCommands

Commit:
Merge branch 'slash-commands' of https://github.com/Catboi8/eris into slash-commands

Commit:
Fixed "will will" typo

Commit:
Removed unneeded if statements, embed => embeds, fixed docs

Commit:
make interaction.edit messageId docs more clear

JustCat80 committed:
Merge pull request #26 from abalabahaha/dev

Pull dev updates

s4093120 reviewed (changes requested):
Consider changing createFollowup to createReply.

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into abalabahaha-dev

Commit:
Merge branch 'abalabahaha-dev' into interactions

JustCat80 committed:
Merge pull request #30 from abalabahaha/dev

Dev updates

Commit:
docs improvements

Constants.InteractionResponseType => Constants.InteractionResponseTypes
Moved ts docs to the same spot
Added more missing ts docs
Fixed some js docs

Added "Errors with ephemeral messages." to editParent

Fix some docs

Fix and improve ts docs
Fix createInteractionResponse docs

Commit:
Ignore .vscode

Commit:
Fixed some defer docs, acknowledge now has a reason to exist

Acknowledge now works differently depending on the interaction type,
Works as a Pong when used on a type 1 interaction. (Ping)
Works as defer() when used on a type 2 interaction. (Slash Command)
Works as deferUpdate() when used on a type 3 interaction. (Message Comonent)

Commit:
Add Donovan's suggested changes, also values is now an array of strings

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

reinacchi commented on index.d.ts:
```diff
@@ -68,7 +69,33 @@ declare namespace Eris {
+  type InteractionContent = {
+    allowedMentions?: AllowedMentions;
+    content?: string;
+    embeds?: EmbedOptions;
```
```suggestion
    embeds?: EmbedOptions[];
```
Should be an array of [embeds](https://discord.com/developers/docs/resources/channel#embed-object) instead

[Here](https://discord.com/developers/docs/interactions/slash-commands#interaction-response-object-interaction-application-command-callback-data-structure)

JustCat80 replied:
yep i missed this

Commit:
embeds is now an actual array of EmbedOptions

Co-Authored-By: Reinhardt <notmarx.tech@gmail.com>

Commit:
Remove Guild#deleteGuildCommand, fix some docs

Commit:
Linted ts

Commit:
Remove unrelated endpoints

Commit:
Added interaction.getOriginal() https://discord.com/developers/docs/interactions/slash-commands#get-original-interaction-response

Commit:
change function names to make more sense, ts docs

getOriginal => getOriginalMessage
edit => editMessage
delete => deleteMessage

Commit:
"Warning: Will error with ephemeral messages." is now put below the notes

Commit:
slight jsdoc improvements

flags was a Boolean in the jsdocs, when it is a Number
slightly change wording

Commit:
(see description) Added deleteOriginalMessage and editOriginalMessage

I will remove these if asked to, I added them because of getOriginalMessage being a thing, let me know if you think these shouldn't stay

Commit:
Id => ID

Commit:
Fix some docs

Commit:
Added Permission Functions to client and guild

- Client#getGuildCommandPermissions
- Client#getCommandPermissions
- Client#editCommandPermissions
- Client#bulkEditCommandPermissions

- Guilld#getGuildCommandPermissions
- Guilld#getCommandPermissions
- Guilld#editCommandPermissions
- Guilld#bulkEditCommandPermissions

remove trailing spaces

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+* @prop {Array<Object>?} data.options.options The run Slash Command options (Mutually exclusive with value)
+* @prop {String?} guildID The id of the guild in which the interaction was created
+* @prop {String} id The interaction id
+* @prop {Member?} member the member who triggered the interaction (This is only sene when the interaction is invoked within a guild)
```
```suggestion
* @prop {Member?} member The member who triggered the interaction (This is only sent when the interaction is invoked within a guild)
```
A typo? Should be `sent`

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+
+/**
+* Represents an interaction
+* @prop {String} applicationID The id of the interaction's application
```
```suggestion
* @prop {String} applicationID The ID of the interaction's application
```
`id` -> `ID`

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+* @prop {Number} data.options.type Command option type, 1-9
+* @prop {Number?} data.options.value The value of the run Slash Command (Mutually exclusive with options)
+* @prop {Array<Object>?} data.options.options The run Slash Command options (Mutually exclusive with value)
+* @prop {String?} guildID The id of the guild in which the interaction was created
```
```suggestion
* @prop {String?} guildID The ID of the guild in which the interaction was created
```
`id` -> `ID`

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+* @prop {Number?} data.options.value The value of the run Slash Command (Mutually exclusive with options)
+* @prop {Array<Object>?} data.options.options The run Slash Command options (Mutually exclusive with value)
+* @prop {String?} guildID The id of the guild in which the interaction was created
+* @prop {String} id The interaction id
```
```suggestion
* @prop {String} id The ID of the interaction
```

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+/**
+* Represents an interaction
+* @prop {String} applicationID The id of the interaction's application
+* @prop {String?} channelID The interaction token
```
```suggestion
* @prop {String?} channelID The ID of the channel in which the interaction was created
```
Sounds like a typo to me.

reinacchi commented on lib/structures/Interaction.js:
```diff
@@ -0,0 +1,328 @@
+* @prop {Message?} message The message the interaction came from (Message Component only)
+* @prop {String} token The interaction token (Interaction tokens are valid for 15 minutes after initial response and can be used to send followup messages but you must send an initial response within 3 seconds of receiving the event. If the 3 second deadline is exceeded, the token will be invalidated.)
+* @prop {Number} type 1 is a Ping, 2 is a Slash Command, 3 is a Message Component
+* @prop {User?} user the user who triggered the interaction (This is only sent when the interaction is invoked within a dm)
```
```suggestion
* @prop {User?} user The user who triggered the interaction (This is only sent when the interaction is invoked within a dm)
```

Commit:
Added suggested docs changes

Co-Authored-By: Reinhardt <notmarx.tech@gmail.com>

Commit:
Added CommandTypes to Constants and command.type to jsdocs

Note: Currently in private beta

Commit:
Added ts docs :)

Will need to be double checked, I'm not the greatest at ts

Commit:
Oops, linted ts

Commit:
Added NUMBER (type 10) to Constants.CommandOptionTypes

JustCat80 committed:
Update index.d.ts

Co-authored-by: DetachHead <57028336+DetachHead@users.noreply.github.com>

Commit:
Added new Interaction types that extend Interaction, added ts docs, changed Constants, added some missing js and tsdocs

PingInteraction (type 1)
CommandInteraction (type 2)
ComponentInteraction (type 3)
UnknownInteraction (any types in the future eris doesn't support yet)

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2250,94 @@ declare namespace Eris {
+  export class CommandInteraction extends Interaction {
+    type: 2;
+    channelID: string;
+    data?: {
```
data is always sent, as per docs, so not optional

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2250,94 @@ declare namespace Eris {
+  export class ComponentInteraction extends Interaction {
+    channelID: string;
+    data?: {
+      componentType?: 2 | 3;
```
Same as last, data isn't optional

DonovanDMC replied:
I also don't think componentType and values can be missing on component interactions, but I don't have a source for that

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2250,94 @@ declare namespace Eris {
+
+  export class UnknownInteraction extends Interaction {
+    channelID?: string;
+    data?: object;
```
This should be unknown, since we don't know what it is

Commit:
fix ts docs

data isn't optional for CommandInteraction and ComponentInteraction
componentType isn't optional for ComponentInteraction
data is unknown for UnknownInteraction
type is numberfor UnknownInteraction

Commit:
for...in => for...of

Co-Authored-By: DetachHead <57028336+DetachHead@users.noreply.github.com>

Commit:
Merge branch 'slash-commands' into combine

Commit:
Merge branch 'combine' into interactions

Commit:
Forgot Linting

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2333,94 @@ declare namespace Eris {
+    channelID: string;
+    data: {
+      componentType: 2 | 3;
+      custom_id: string;
```
this should be camelCased to `customID`, like the other properties around it

JustCat80 replied:
Right now it isn't because #1189 also does not, i will if bsian does aswell as if not it would be confusing to users

bsian03 replied:
The reason why my components PR doesn't camelcase is because it would be an absolute nightmare trying to map every single component type to the correct value. I haven't checked what's going on here but it would be preferable if it was kept consistent with my PR.

JustCat80 replied:
I figured, more work and it wouldnt work for any new components till updated

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2333,94 @@ declare namespace Eris {
+    data: {
+      componentType: 2 | 3;
+      custom_id: string;
+      values?: string[];
```
it might be worth making button & select menu interactions distinct *in types only*, so we don't have to check if this exists, componentType can be used to narrow (though this might be more of a mess than it's worth)

JustCat80 replied:
yeaa I honestly think it is, and exists with select menus so just check the component type?

DonovanDMC commented on index.d.ts:
```diff
@@ -2207,6 +2333,94 @@ declare namespace Eris {
+    data: {
+      id: string;
+      name: string;
+      resolved?: {
```
these would look better using `Record`, e.g. `Record<string, User>`

DonovanDMC commented on index.d.ts:
```diff
@@ -68,7 +69,71 @@ declare namespace Eris {
+    name: string;
+    description: string;
+    required?: boolean;
+    choices?: { name: string; value: string | number};
```
choices is an array
```suggestion
    choices?: { name: string; value: string | number }[];
```

DonovanDMC replied:
this should also have an options property, as top level options can have nested options (sub commands)

DonovanDMC reviewed (changes requested):
I'm going to get banned for gh spam

bsian got onto me so I'm going to start adding edits onto this comment

* you should use `Constants["key"][keyof Constants["key"]]` for number unions instead of completely redefining the union, ex `Constants["CommandOptionTypes"][keyof Constants["CommandOptionTypes"]]` instead of remaking the union type
* and you didn't add the classes to the exports, ex DonovanDMC/eris@a64bad01b96e973757be321daf87a6e60a691d5c
* on the `Interaction` type, the union `type: InteractionType | number;` does nothing, it should just be made `number` (as it gets converted to that)

Commit:
Added requested changes

now use keyof
added classes to exports
changed interaction type to number

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Added more requested changes

now use Record
componentType is now component_type
choices is an array

DonovanDMC commented on lib/structures/CommandInteraction.js:
```diff
@@ -0,0 +1,234 @@
+                content.content = "" + content.content;
+            }
+        }
+        return this._client.executeWebhook.call(this._client, this.applicationID, this.token, content);
```
This should either be converted into its own client function, or wait should be added so executeWebhook actually constructs the message (it doesn't attempt it unless await is true)

(I think we use Object.assign instead of spread around here)
```suggestion
        return this._client.executeWebhook.call(this._client, this.applicationID, this.token, Object.assign({ wait: true }, content);
```

Commit:
Assign wait to all createFollowup methods

DonovanDMC commented on index.d.ts:
```diff
@@ -68,7 +69,58 @@ declare namespace Eris {
+  type InteractionDataOptions = {
+    name: string;
+    type: Constants["CommandOptionTypes"][keyof Constants["CommandOptionTypes"]];
+    value?: Constants["CommandOptionTypes"][keyof Constants["CommandOptionTypes"]];
```
`value` should be string
```suggestion
    value?: string;
```

JustCat80 replied:
*wait yea why did I change that*

JustCat80 committed:
Change value back to string

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Fix Guild#bulkEditCommands

DonovanDMC commented on lib/structures/ComponentInteraction.js:
```diff
@@ -0,0 +1,275 @@
+            }
+        }
+        return this._client.createInteractionResponse.call(this._client, this.id, this.token, {
+            type: InteractionResponseTypes.CHANNEL_MESSAGE_WITH_SOURCE,
```
This is the wrong interaction response type for components
```suggestion
            type: InteractionResponseTypes.UPDATE_MESSAGE,
```

DonovanDMC replied:
I forgot to check the other methods, but it looks `defer` is also using the slash command variant

JustCat80 replied:
Defer is the same as the slash commands one, deferUpdate is not, I just forgot to change acknowledge back to what it was

JustCat80 replied:
Also I think the method was defer_update_message

JustCat80 replied:
Wait no that's correct I think, component interactions use the same responses, but have a few extra, aka editParent and deferUpdate

JustCat80 replied:
Just checked, it is already correct

JustCat80 committed:
Merge pull request #46 from Catboi8/interactions

Interactions updated

Commit:
Added Interaction.acknowledged, update Interaction.acknowledged when using a function that a acknowledges the interaction

Commit:
acknowledged docs

JustCat80 committed:
Push Acknowledged consept to main interactions branch

Push Acknowledged consept to main interactions branch

Commit:
Lint ts

DonovanDMC commented on lib/Client.js:
```diff
@@ -569,6 +664,28 @@ class Client extends EventEmitter {
+    * @returns {Promise}
+    */
+    createInteractionResponse(interactionID, interactionToken, options) {
+        return this.requestHandler.request("POST", Endpoints.INTERACTION_RESPOND(interactionID, interactionToken), true, options);
```
To stop the 3 line+ log spam these cause, this should be added
```suggestion
        return this.requestHandler.request("POST", Endpoints.INTERACTION_RESPOND(interactionID, interactionToken), true, options, undefined, "/interactions/:id/:token/callback");
```
(also can you not send files through interactions? this doesn't have a way to provide them)

JustCat80 replied:
You cannot yet, tho you can after the initial response

DonovanDMC replied:
ah, no wonder I've never had an issue, I've only done them in followups

JustCat80 committed:
Update lib/Client.js

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Added file to createMessage docs because after the initial response you can send a file

Commit:
Docs changes, Constants changes

Change mentions of Slash Command to Application Command unless directly referencing the Slash Command Application Command type

Loliticos commented on lib/Constants.js:
```diff
@@ -274,3 +273,45 @@ module.exports.Intents = {
+};
+
+module.exports.CommandTypes = {
+    COMMAND: 1,
```
This should probally be changed to `CHAT_INPUT` in order to match the documentation

JustCat80 replied:
ngl chat_input is such a stupid name and I am very salty but if a few people agree I will change it

bsian03 replied:
Blame Discord

Commit:
Changed Constants COMMAND to CHAT_INPUT, changed name check

reinacchi reviewed (changes requested):
[`useSlashCommands`](https://github.com/Catboi8/eris/blob/4415d02f3d5b6e4138c862804c21b63112ef885f/lib/Constants.js#L68) -> `useApplicationCommands` to match with what is in the [docs](https://discord.com/developers/docs/topics/permissions#permissions-bitwise-permission-flags)

![image_2021-08-11_211044](https://user-images.githubusercontent.com/73813638/129044505-432a4dd4-1edd-4566-bc9e-481e6debd111.png)

TTtie committed:
use raw messages when interaction messages are ephemeral

TTtie committed:
use a bit mask instead of direct comparison

TTtie committed:
use an explicit 64 for PossiblyEphemeralMessage.flags

JustCat80 committed:
Merge pull request #52 from TTtie/ephemeral-message-passthrough

Commit:
Removed InteractionFlags in favor of adding EPHEMERAL to MessageFlags, fixed some constants styling

Commit:
Application Commands Example

Loliticos commented on examples/applicationCommands.js:
```diff
@@ -0,0 +1,103 @@
+        bot.createCommand({
+            name: "Test Message Menu",
+            type: Constants.CommandTypes.MESSAGE
+        }); //Create a user context menu
```
I guess this was meant to be message right?
```suggestion
        }); //Create a message context menu
```

JustCat80 replied:
Oop yea I'm stupid, when I changed the wording I forgot to change that

JustCat80 committed:
Fix message context menu comment

Co-authored-by: Loliticos <47262825+Loliticos@users.noreply.github.com>

Commit:
editOriginalMessage doesnt error with ephemeral messages

i thought it did but maybe im dumb

Commit:
Warn if no intents instead of allowing it to error, as Interactions dont need an intent

Emit a warning and set to intents 0 if none are specified

Commit:
Change applicationCommands example

Commit:
Oops added an underscore to BOT TOKEN

Commit:
Update Constants Names to be consistent

Commit:
Added docs for forgotten context menu data

Commit:
remove PossiblyEphemeralMessage from ts docs

Commit:
Fix ComponentInteraction#message docs, make most message methods check if the message is ephemeral before attempting the method

Commit:
CONTEXT_MENU_COMMAND message type

Commit:
APPLICATION_COMMAND => CHAT_INPUT_COMMAND

Add forgotten `case MessageTypes.CONTEXT_MENU_COMMAND` to Message.js

Commit:
data => info in interaction classes, data.resolved fixed

Commit:
Oops, fix redefining things for no reason

Commit:
remove data.resolved for now

Commit:
remove unused things for now

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into interactions

Commit:
Add components to jsdocs

Commit:
interaction.data.resolved is now actually set instead of just passing the raw data

Commit:
Fix PartialChannel and PartialRole having id set as number

Commit:
fixed docs to reflect that data.resolved are Collections

Commit:
add APPLICATION_COMMAND events, will maybe add an ApplicationCommand class later

DonovanDMC commented on index.d.ts:
```diff
@@ -536,6 +589,10 @@ declare namespace Eris {
+    (event: "interactionCreate", listener: (interaction: PingInteraction | CommandInteraction | ComponentInteraction | UnknownInteraction) => void): T;
+    (event: "commandCreate", listener: (interaction: ApplicationCommand) => void): T;
+    (event: "commandUpdate", listener: (interaction: ApplicationCommand) => void): T;
+    (event: "commandDelete", listener: (interaction: ApplicationCommand) => void): T;
```
```suggestion
    (event: "commandCreate", listener: (command: ApplicationCommand) => void): T;
    (event: "commandUpdate", listener: (command: ApplicationCommand) => void): T;
    (event: "commandDelete", listener: (command: ApplicationCommand) => void): T;
```

not really important, but the parameter name should reflect what its value is

JustCat80 replied:
oop I had forgotten to do this, doesn't matter now

Commit:
Revert commit cbe15029c223b20f73ae1a07905d9769f0126819 as these events are not sent to bots

Revert "add APPLICATION_COMMAND events, will maybe add an ApplicationCommand class later"
This reverts commit cbe15029c223b20f73ae1a07905d9769f0126819.

Commit:
fix message TypeError (hotfix by GweeKwee)

fix TypeError: Cannot read property 'members' of undefined at new Message

Co-Authored-By: Gwee&Kwee <gwee@nekotina.com>

Commit:
interaction.channelID is now interaction.channel, similar to message

Commit:
fix up message.interaction, make message use this._client so its consistent

Commit:
Merge branch 'dev' into interactions

JustCat80 commented on lib/structures/CommandInteraction.js:
```diff
@@ -0,0 +1,382 @@
+                const membermap = new Collection(Member);
+                Object.entries(info.data.resolved.members).forEach(([id, member]) => {
+                    member.user = {id};
+                    membermap.set(id, new Member(member, this._client.guilds.get(this.guildID), this._client));
```
oop I am stupid

Commit:
use info.guild_id when defining members

Commit:
Update UnknownInteraction.message

Commit:
Cache CommandInteraction.resolved users and members, cache interaction.member and interaction.user

coolcalcacol commented on index.d.ts:
```diff
@@ -73,7 +74,58 @@ declare namespace Eris {
+  type InteractionDataOptions = {
+    name: string;
+    type: Constants["ApplicationCommandOptionTypes"][keyof Constants["ApplicationCommandOptionTypes"]];
+    value?: string;
```
Is this supposed to be a string or the type from line 81?

JustCat80 replied:
90% sure it's the type

JustCat80 replied:
yep https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-interaction-data-option-structure

JustCat80 replied:
actually it says the value, but links to the type, im not sure, its mutually exclusive with options tho so i have no idea

coolcalcacol replied:
perhaps just change it to this?
```suggestion
    value?: string | number | boolean;
```

JustCat80 replied:
Yea ill make it that until we prove otherwise

Commit:
Change options.value to string | number | boolean for now

Co-Authored-By: coolcalcacol <31094492+coolcalcacol@users.noreply.github.com>

DonovanDMC committed:
first attempt

DonovanDMC committed:
add stuff to unions

JustCat80 committed:
Merge pull request #61 from DonovanDMC/option-types

make InteractionDataOptions strictly defined

Commit:
Convert to interfaces

DonovanDMC commented on index.d.ts:
```diff
@@ -73,7 +74,77 @@ declare namespace Eris {
+  }
+
+  type InteractionContent = Pick<WebhookPayload, "content" | "embeds" | "allowedMentions" | "tts" | "flags">;
+  type InteractionWebhookContent = Pick<WebhookPayload, "content" | "embeds" | "file" | "allowedMentions" | "tts" | "flags">;
```
```suggestion
  type InteractionContent = Pick<WebhookPayload, "content" | "embeds" | "allowedMentions" | "tts" | "flags" | "components">;
  type InteractionWebhookContent = Pick<WebhookPayload, "content" | "embeds" | "file" | "allowedMentions" | "tts" | "flags" | "components">;
```

This should be added since components are now in dev

JustCat80 replied:
Oop yep i forgot to update the ts

JustCat80 committed:
Add components to webhook content in ts

Co-authored-by: Donovan Daniels <hewwo@yiff.rocks>

Commit:
Add Interaction classes to esm.mjs

Commit:
Add ComponentTypes to constants because bsian didn't

:)

Co-Authored-By: Donovan Daniels <hewwo@yiff.rocks>

DonovanDMC committed:
Even More Detailed Interaction Types (#63)

* ComponentTypes constants

* ComponentInteraction*Data types

* use constants for type

* space

* More specific typings for `ApplicationCommandOptions`

* forgot to change stuff when I copied

* fixes and changes

* even more messy types

* correct choices types

* lint

* specific depth limits
docs: https://discord.com/developers/docs/interactions/application-commands#subcommands-and-subcommand-groups

* `ApplicationCommand#description` ternary

* these have been resolved

* lint

* even more

Commit:
ts hotfix

Commit:
remove UnknownInteraction

Commit:
Undo last commit

Commit:
export Interaction

Commit:
data.interaction.guild_id => data.guild_id in Message

Commit:
Can send files on initial interaction response, add .then update to UnknownInteraction methods

Question life because why is executeWebhook file different from createMessage file wtf

DonovanDMC committed:
More Fixes (#64)

* I suck

* same thing here

Commit:
Fix Error: Only iterables may be used in a SetType (Catboi8#68)

Fix allowed_mentions in interaction responses

Co-Authored-By: Gwee&Kwee <gwee@nekotina.com>

Commit:
Added ephemeral to Attachment ts

Commit:
Fix editParent typings

cillynder commented on lib/Client.js:
```diff
@@ -339,6 +343,60 @@ class Client extends EventEmitter {
+            throw new Error("You must provide an id of the guild whose permissions you want to edit.");
+        }
+
+        return this.requestHandler.request("PUT", Endpoints.GUILD_COMMAND_PERMISSIONS(this.application.id, guildID), true, {permissions});
```
```suggestion
        return this.requestHandler.request("PUT", Endpoints.GUILD_COMMAND_PERMISSIONS(this.application.id, guildID), true, permissions);
```
This doesn't seem to work as Discord expects an [array of partial permissions](https://discord.com/developers/docs/interactions/application-commands#batch-edit-application-command-permissions) for bulk edits
Removing the curly brackets around `permissions` should fix it

cillynder commented on index.d.ts:
```diff
@@ -1561,6 +1713,9 @@ declare namespace Eris {
     addRelationship(userID: string, block?: boolean): Promise<void>;
     addSelfPremiumSubscription(token: string, plan: string): Promise<void>;
     banGuildMember(guildID: string, userID: string, deleteMessageDays?: number, reason?: string): Promise<void>;
+    bulkEditCommandPermissions(guildID: string, permissions: ApplicationCommandPermissions[]): Promise<GuildApplicationCommandPermissions[]>;
```
```suggestion
    bulkEditCommandPermissions(guildID: string, permissions: { id: string; permissions: ApplicationCommandPermissions }[]): Promise<GuildApplicationCommandPermissions[]>;
```
`permissions` type is wrong
Not sure if this should be separated into another type; I think it's short enough though.

cillynder commented on index.d.ts:
```diff
@@ -73,7 +74,115 @@ declare namespace Eris {
+  interface InteractionDataOptionsSubCommand {
+    name: string;
+    type: Constants["ApplicationCommandOptionTypes"]["SUB_COMMAND"];
+    options: InteractionDataOptions[];
```
```suggestion
    options?: InteractionDataOptions[];
```
The [docs seem to be a bit confusing about this](https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-interaction-data-option-structure), but `options` appears to be optional in my testing. If the subcommand doesn't have any options, `options` here won't exist. This is also the case if the subcommand does accept non-required options, but none of them are provided in the interaction.

Basically: `options` can't be empty. It's either undefined, or it has something in it.

JustCat80 committed:
Fix bulkEditCommandPermissions typings

Co-authored-by: Lava <service.github@lava.moe>

Commit:
Add ts fixes, add channel_types to ApplicationCommandOptionsChannel

Co-Authored-By: Lava <me@lava.moe>

Commit:
Merge branch 'interactions' of https://github.com/Catboi8/eris into interactions

Commit:
Oops

Commit:
fix perms again xD

Co-Authored-By: Lava <me@lava.moe>

Commit:
change all acknowledges, make PingInteraction consistent

Commit:
Autocomplete Interaction Support 🥳

Add support for interaction type APPLICATION_COMMAND_AUTOCOMPLETE (4)
add unknown to ComponentInteraction data
fix InteractionContent
add acknowledge to UnknownInteraction (is just createInteractionResponse)

Commit:
Merge branch 'dev' of https://github.com/abalabahaha/eris into interactions

Commit:
Fix js docs maybe

DonovanDMC commented on index.d.ts:
```diff
@@ -2506,15 +2521,15 @@ declare namespace Eris {
   export class ComponentInteraction<T extends PossiblyUncachedTextable = TextableChannel> extends Interaction {
     channel: T;
-    data: ComponentInteractionButtonData | ComponentInteractionSelectMenuData;
+    data: ComponentInteractionButtonData | ComponentInteractionSelectMenuData | unknown;
```
Why was unknown added here?
(fix added into Catboi8/eris#69)

DonovanDMC replied:
This makes using the property impossible without explicitly casting to one of the two

![](https://butts-are.cool/Code_-_Insiders_09-26-2021_14-36-55.png)

DonovanDMC committed:
a few fixes/improvements (#69)

* start

* remove this unknown garbage

Awoocado commented on lib/Constants.js:
```diff
@@ -279,3 +280,49 @@ module.exports.Intents = {
+    ACTION_ROW: 1,
+    BUTTON: 2,
+    SELECT_MENU: 3
+};
```
```suggestion
module.exports.ComponentTypes = {
    ACTION_ROW:  1,
    BUTTON:      2,
    SELECT_MENU: 3
};
```

JustCat80 committed:
ComponentTypes Lint

Co-authored-by: Avocado <43632131+Awoocado@users.noreply.github.com>

reinacchi commented on index.d.ts:
```diff
@@ -60,7 +61,145 @@ declare namespace Eris {
+
+  type AnyApplicationCommand = ChatInputApplicationCommand | MessageApplicationCommand | UserApplicationCommand;
+  type ChatInputApplicationCommand = ApplicationCommand<Constants["ApplicationCommandTypes"]["CHAT_INPUT"]>;
+  type MessageApplicationCommand = Omit<ApplicationCommand<Constants["ApplicationCommandTypes"]["USER"]>, "description" | "options">;
```
```suggestion
  type UserApplicationCommand = Omit<ApplicationCommand<Constants["ApplicationCommandTypes"]["USER"]>, "description" | "options">;
```

reinacchi commented on index.d.ts:
```diff
@@ -60,7 +61,145 @@ declare namespace Eris {
+  type AnyApplicationCommand = ChatInputApplicationCommand | MessageApplicationCommand | UserApplicationCommand;
+  type ChatInputApplicationCommand = ApplicationCommand<Constants["ApplicationCommandTypes"]["CHAT_INPUT"]>;
+  type MessageApplicationCommand = Omit<ApplicationCommand<Constants["ApplicationCommandTypes"]["USER"]>, "description" | "options">;
+  type UserApplicationCommand = Omit<ApplicationCommand<Constants["ApplicationCommandTypes"]["MESSAGE"]>, "description" | "options">;
```
```suggestion
  type MessageApplicationCommand = Omit<ApplicationCommand<Constants["ApplicationCommandTypes"]["MESSAGE"]>, "description" | "options">;
```

JustCat80 replied:
@DonovanDMC 10/10

JustCat80 replied:
~~god damn it can you resuggest changes it won't let me commit them now~~

nvm

JustCat80 committed:
Update index.d.ts

Co-authored-by: Reinhardt <notmarx.tech@gmail.com>

JustCat80 committed:
Update index.d.ts

Co-authored-by: Reinhardt <notmarx.tech@gmail.com>

JustCat80 committed:
Add ButtonStyles to Constants

*glares at bsian*

JustCat80 committed:
Components example

Because explaining it is meh, direct em here instead :3

JustCat80 committed:
Oops forgot intents
